### PR TITLE
Heuristics Interface

### PR DESCRIPTION
**Pull request by dhendryc**
Interface for our own predefined and user custom heuristics.

### CODE REVIEW

**Comment by dhendryc** [comment]
We should also think about whether the heuristics should be called at every node or if this should depend on the depth in the tree etc.

File: src/heuristics.jl [hunk 1]
```diff
@@ -2,3 +2,78 @@
...
+`h` is the heuristic function receiving as input the tree, the bounded LMO and a point x (the current node solution).
+It returns the heuristic solution (can be nothing, we check for that) and whether feasibility still has to be check.
+`prob` is the probability with which it will be called.        
+"""
+# Would 'Heuristic' also suffice? Or might we run into Identifer conflicts with other packages?
+struct Heuristic
+    h::Function
+    prob::Float64
+    identifer::Symbol
+end
```
Comment (matbesancon) [inline src/heuristics.jl:18]:
```suggestion
struct Heuristic{F}
    h::F
    prob::Float64
    identifer::Symbol
end
```
to have it type-stable

File: src/heuristics.jl [hunk 2]
```diff
@@ -2,3 +2,78 @@
...
+Flip coin.
+"""
+function flip_coin(w=0.5)
+    return rand() ≤ w
+end
+
+"""
+Add a new solution found from the heuristic to the tree.
+"""
+function add_heuristic_solution(tree, x, val, heu::Symbol)
```
Comment (matbesancon) [inline src/heuristics.jl:33]:
```suggestion
function add_heuristic_solution(tree, x, val, heuristic_name::Symbol)
```

better use complete variable names

File: src/heuristics.jl [hunk 3]
```diff
@@ -2,3 +2,78 @@
...
+    end
+    tree.incumbent = val
+    Bonobo.bound!(tree, node.id)
+end
+
+"""
+Choose which heuristics to run by rolling a dice.
+"""
+# TO DO: We might want to change the probability depending on the depth of the tree
+# or have other additional criteria on whether to run a heuristic
```
Comment (matbesancon) [inline src/heuristics.jl:50]: docstring must be attached to the function, without comment in-between

File: src/interface.jl [hunk 4]
```diff
@@ -51,6 +51,7 @@ fw_verbose             - if true, FrankWolfe logs are printed
 use_shadow_set         - The shadow set is the set of discarded vertices which is inherited by the children nodes.
                         It is used to avoid recomputing of vertices in case the LMO is expensive. In case of a cheap LMO,
                         performance might improve by disabling this option. 
+custom_heuristics      - List of  custom heuristic from the user. The default is a dummy heuristic that return nothing and is, in practice, never called.
```
Comment (matbesancon) [inline src/interface.jl:54]:
```suggestion
custom_heuristics      - List of  custom heuristic from the user.
```

no need to specify the implementation details

File: src/heuristics.jl (hunk 1 above)
Comment (matbesancon) [reply src/heuristics.jl:18]: also, maybe `run_heuristic` instead of `h`

File: src/polytope_blmos.jl [hunk 5]
```diff
@@ -118,3 +168,21 @@ function is_simple_linear_feasible(sblmo::UnitSimplexSimpleBLMO, v)
     end
     return sum(v) ≤ sblmo.N + 1e-3
 end
+
+"""
+Hyperplane aware rounding for the Unit simplex.
```
Comment (matbesancon) [inline src/polytope_blmos.jl:173]:
```suggestion
Hyperplane-aware rounding for the Unit simplex.
```

File: test/LMO_test.jl [hunk 6]
```diff
@@ -166,3 +166,50 @@ diffi = x_sol + 0.3*rand([-1,1], n)
...
+
+    x, _, result =
+        Boscia.solve(f, grad!, sblmo, fill(0.0, n), fill(1.0*N, n), collect(1:n), n, custom_heuristics=[heu])
+
+    @test sum(isapprox.(x, x_sol, atol=1e-6, rtol=1e-2)) == n
+    @test isapprox(f(x), f(result[:raw_solution]), atol=1e-6, rtol=1e-3)
+end
+
+n = 20
+x_sol = rand(1:Int(floor(n/4)), n)
```
Comment (matbesancon) [inline test/LMO_test.jl:195]:
```suggestion
x_sol = rand(1:floor(Int, n/4), n)
```

Comment (matbesancon) [reply test/LMO_test.jl:195]: specifies to the floor function that it must return an integer

Comment (matbesancon) [reply test/LMO_test.jl:195]: to apply everywhere

**Comment by pokutta** [comment]
just went over this -> it would be good if a heuristics could return more than one solution as they often create multiple sols. maybe they can return a list or something instead of a single solution.

**Comment by dhendryc** [comment]
Count the calls to the LMO in the heuristcs:
1. Wrap the inner LMO in a TrackingLMO just for the heuristics:
```
lmo = tracking_lmo.inner
heuristic_lmo = TrackingLMO()
run_heuristic(heuristic_lmo)
collect_statistics(heuristic_lmo)
```
and have a field specifically for the heuristic LMO calls in the TrackingLMO.

**Comment by pokutta** [comment]
- bumped to prob 1.0 for standard rounding
- added two simple heuristics -> not perfect yet but a start
- adjusted two examples to use the heuristics (nonlinear.jl and portfolio.jl)
- cleaned up some imports

File: src/heuristics.jl [hunk 7]
```diff
@@ -2,3 +2,126 @@
...
+
+# Create default heuristic. Doesn't do anything and should be called.
+Heuristic() = Heuristic(x -> nothing, -0.1, :default)
+
+"""
+Flip coin.
+"""
+function flip_coin(w=0.5)
+    return rand() ≤ w
+end
```
Comment (matbesancon) [inline src/heuristics.jl:28]: this makes the function deterministic by passing an RNG

```suggestion
function flip_coin(w=0.5, rng=Random.GLOBAL_RNG)
    return rand(rng) ≤ w
end
```

File: src/heuristics.jl [hunk 8]
```diff
@@ -2,3 +2,126 @@
...
+        x_new = Boscia.compute_extreme_point(blmo, nabla)
+        push!(sols, x_new)
+    end
+    return sols, false
+end
+
+
+"""
+Advanced lmo-aware rounding for binary vars
+"""
```
Comment (matbesancon) [inline src/heuristics.jl:114]: docstrings need to be attached to the function to be a docstring one has access to

File: src/heuristics.jl [hunk 9]
```diff
@@ -2,3 +2,126 @@
...
+    for idx in tree.branching_indices
+        x_rounded[idx] = round(x[idx])
+    end
+    return [x_rounded], true
+end
+
+
+"""
+Follow-the-gradient heuristic.
+"""
```
Comment (matbesancon) [inline src/heuristics.jl:93]: same here

**Comment by dhendryc** [comment]
The feasibility check after the rounding can become expensive. We could have parameters in the `solve` function allowing the user to set the probability for the standard heurstics, i.e. rounding, gradient following and rounding lmo.

**Comment by matbesancon** [comment]
More generically, we can define a property for heuristics, that describes whether they are feasible by design or require a feasibility check, and then a boolean keyword that allows heuristics or not

**Comment by pokutta** [comment]
> The feasibility check after the rounding can become expensive. It would be good to enable the user to set the probability of the rounding to something smaller than 1.0 in case the feasibility check of their LMO is expensive.

you mean of the "standard" rounding -> yes good point. how did you do it before making it a heuristic?

**Comment by pokutta** [comment]
> More generically, we can define a property for heuristics, that describes whether they are feasible by design or require a feasibility check, and then a boolean keyword that allows heuristics or not

we have the flag that we return whether feasiblity needs to be checked or not. i set them to zero for those that are guaranteed to be feasible. do we need more?

**Comment by dhendryc** [comment]
> > More generically, we can define a property for heuristics, that describes whether they are feasible by design or require a feasibility check, and then a boolean keyword that allows heuristics or not
> 
> we have the flag that we return whether feasiblity needs to be checked or not. i set them to zero for those that are guaranteed to be feasible. do we need more?

No, that's fine. Then we only need a property for the "standard" rounding.
And yes, before we rounded every time. I think we didn't have large enough problems yet where the feasibility test was problematic. This showed up for the network design problems with Kartikey.

**Pull request merged**